api/src/main/java/org/pmiops/workbench/api/EgressEventsAdminController.java

```diff
@@ -1,20 +1,105 @@
+import org.pmiops.workbench.exceptions.BadRequestException;
+import org.pmiops.workbench.exceptions.NotFoundException;
 import org.pmiops.workbench.model.Authority;
 import org.pmiops.workbench.model.ListEgressEventsRequest;
 import org.pmiops.workbench.model.ListEgressEventsResponse;
+import org.pmiops.workbench.utils.PaginationToken;
+import org.pmiops.workbench.utils.mappers.EgressEventMapper;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
 import org.springframework.http.ResponseEntity;
 import org.springframework.web.bind.annotation.RestController;
 
 @RestController
 public class EgressEventsAdminController implements EgressEventsAdminApiDelegate {
+  private static final int DEFAULT_PAGE_SIZE = 32;
+  private static final int MAX_PAGE_SIZE = 128;
+
+  @Autowired private EgressEventMapper egressEventMapper;
+  @Autowired private UserDao userDao;
+  @Autowired private WorkspaceDao workspaceDao;
+  @Autowired private EgressEventDao egressEventDao;
 
   @AuthorityRequired(Authority.SECURITY_ADMIN)
   @Override
   public ResponseEntity<ListEgressEventsResponse> listEgressEvents(
       ListEgressEventsRequest request) {
-    throw new NotImplementedException();
+    int pageSize = DEFAULT_PAGE_SIZE;
+    if (request.getPageSize() != null && request.getPageSize().intValue() > 0) {
+      pageSize = Math.min(request.getPageSize().intValue(), MAX_PAGE_SIZE);
+    }
+
+    Pageable pageable = PageRequest.of(0, pageSize);
+    if (!Strings.isNullOrEmpty(request.getPageToken())) {
+      PaginationToken token = PaginationToken.fromBase64(request.getPageToken());
+      if (!token.matchesParameters(toPaginationParams(request))) {
+        throw new BadRequestException("search parameters changed between paginated calls");
+      }
+      pageable = PageRequest.of((int) token.getOffset(), pageSize);
+    }
+
+    DbUser userFilter = null;
+    if (!Strings.isNullOrEmpty((request.getSourceUserEmail()))) {
+      userFilter = userDao.findUserByUsername(request.getSourceUserEmail());
+      if (userFilter == null) {
+        throw new NotFoundException(
+            String.format("user '%s' not found", request.getSourceUserEmail()));
+      }
+    }
+
+    DbWorkspace workspaceFilter = null;
+    if (!Strings.isNullOrEmpty((request.getSourceWorkspaceNamespace()))) {
+      workspaceFilter =
+          workspaceDao
+              .getByNamespace(request.getSourceWorkspaceNamespace())
+              .orElseThrow(
+                  () ->
+                      new NotFoundException(
+                          String.format(
+                              "workspace namespace '%s' not found",
+                              request.getSourceWorkspaceNamespace())));
+    }
+
+    Page<DbEgressEvent> page;
+    if (userFilter == null && workspaceFilter == null) {
```

Comment:
nit: It's bit 'backward' while reading the code. Can it be:
```
if(workspaceFilter != null && userFilter != null) {
 findAllByUserAndWorkspaceOrderByCreationTimeDesc
} 
else if(workspaceFilter != null) {
  findAllByWorkspace
} else if(userFilter != null) {
 findByUser
} else {
 findAll? 
```
}

Reply:
Good call, much cleaner